### ISSUE TITLE
Type narrowing is working inconsistently

### ISSUE THREAD

**Issue by urielvan** [microsoft/TypeScript#45707]
# Bug Report

### 🔎 Search Terms

type narrowing, conditional type

### 🕗 Version & Regression Information

v4.4.2

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4BYAKCuADsYkpN0k4BVAZ0bgG8BfKjXqNmaVgAUcYDgB4AKnCQAPBrQAmHOACMIEADapacALxxmergD5eVOHBABXPTGBgDAfgBccORUp2IWgBhAAsUWgBzJC8fRRUkdU0YKAckWzs4dzgACgA3FD1U705GAG0AXQBKE2tciGA1dLtvPIKi9i4oauNa+rU-AWpKTAdaDGBAjsYAZSQDDHk41Q1tXQNwkzMCq2ywKQ5vSQhpeUtqnnSiGAcoI1onPQHBSjRAjixdTez0mRKoWfmMCacECoXCUWMPHyhVYPRs-gydletA4+iQADo9BAIq0YZVgXxBnYAPSWKiVUhAA)

### 🙁 Actual behavior

The error is reported said that `Parameter 'value' implicitly has an 'any' type`

### 🙂 Expected behavior

No error is reported, and the parameter 'value' is considered as type `User` correctly
Btw, if you hover on the property `onChange` at line 18, you can see it narrowed the type correctly

**Comment by ahejlsberg** [comment]
In this declaration

```ts
interface Props<T extends boolean = false> {
  multiple?: T;
  onChange?: T extends true
    ? (value: User[]) => void
    : (value: User) => void;
}
```

the compiler's type inference process doesn't "see" that the conditional type evaluates to a function type in either case. If you change the declaration to

```ts
interface Props<T extends boolean = false> {
  multiple?: T;
  onChange?: (value: T extends true ? User[] : User) => void
}
```

the issue goes away.

**Comment by urielvan** [comment]
@ahejlsberg thanks for explanation
